File: parenthesisvalidation/ParenthesisValidation.java
package parenthesisvalidation;

import springcollection.LinkedStack;

public class ParenthesisValidation {

    public boolean isValid(String string) {
        LinkedStack<Character> chars = new LinkedStack<>();
        for (var i = 0; i < string.length(); i++) {
            var character = string.charAt(i);
            if (character == '(') {
                chars.push(character);
            }
            if (character == ')') {
                if (chars.size() > 0) {
                    chars.pop();
                    continue;
                }
                return false;
            }
        }
        return chars.size() == 0;
    }

    public int longestValidParenthesisSequence(String string) {
        int count = 0;
        LinkedStack<Character> chars = new LinkedStack<>();
        for (var i = 0; i < string.length(); i++) {
            var character = string.charAt(i);
            if (character == '(') {
                chars.push(character);
                count++;
            }
            if (character == ')' && chars.size() > 0) {
                chars.pop();
                count++;
            }
        }
        return count - chars.size();
    }
}

File: parenthesisvalidation/parenthesisvalidation.py
class Solution(object):
    def isValid(self, s: str):
        pList = []
        pPair = {"{": "}", "[": "]", "(": ")"}
        for i in range(0, len(s)):
            char = s[i]
            if char in pPair.keys():
                pList.append(char)
            if char in pPair.values():
                if len(pList) > 0 and char == pPair.get(pList[len(pList) - 1]):
                    pList.pop()
                    continue
                return False
        return len(pList) == 0

print(Solution().isValid('{[}]'))
